ports/esp32/machine_pwm.c

```diff
@@ -128,61 +267,72 @@ STATIC void mp_machine_pwm_init_helper(machine_pwm_obj_t *self,
     mp_arg_parse_all(n_args, pos_args, kw_args,
         MP_ARRAY_SIZE(allowed_args), allowed_args, args);
 
-    int channel;
-    int avail = -1;
+    int channel_idx = find_channel(self->pin, ANY_MODE);
+    if (channel_idx == -1) {
+        mp_raise_msg_varg(&mp_type_ValueError, MP_ERROR_TEXT("out of PWM channels:%d"), PWM_CHANNEL_MAX); // in all modes
+    }
 
-    // Find a free PWM channel, also spot if our pin is
-    //  already mentioned.
-    for (channel = 0; channel < LEDC_CHANNEL_MAX; ++channel) {
-        if (chan_gpio[channel] == self->pin) {
-            break;
+    int freq = args[ARG_freq].u_int;
+    if ((freq < -1) || (freq > 40000000)) {
+        mp_raise_ValueError(MP_ERROR_TEXT("freqency must be between 1 and 40000000 Hz"));
+    }
+    // Check if freq wasn't passed as an argument
+    if (freq == -1) {
+        // Check if already set, otherwise use the default freq.
+        // Possible case:
+        // pwm = PWM(pin, freq=1000, duty=256)
+        // pwm = PWM(pin, duty=128)
+        if (chans[channel_idx].timer_idx != -1) {
+            freq =  timers[chans[channel_idx].timer_idx].freq_hz;
         }
-        if ((avail == -1) && (chan_gpio[channel] == -1)) {
-            avail = channel;
+        if (freq < 0) {
+            freq = PWFREQ;
         }
     }
-    if (channel >= LEDC_CHANNEL_MAX) {
-        if (avail == -1) {
-            mp_raise_ValueError(MP_ERROR_TEXT("out of PWM channels"));
-        }
-        channel = avail;
+
+    int timer_idx = find_timer(freq, SAME_FREQ_OR_FREE, CHANNEL_IDX_TO_MODE(channel_idx));
+    if (timer_idx == -1) {
+        timer_idx = find_timer(freq, SAME_FREQ_OR_FREE, ANY_MODE);
+    }
+    if (timer_idx == -1) {
+        mp_raise_msg_varg(&mp_type_ValueError, MP_ERROR_TEXT("1 out of PWM timers:%d"), PWM_TIMER_MAX); // in all modes
```

Comment:
the "1" in the error text is confusing, I think it should just be "out of PWM timers:%d"